tests/src/test/scala/cats/tests/EitherTTests.scala

```diff
@@ -378,4 +378,33 @@ class EitherTTests extends CatsSuite {
       }
     }
   }
+
+  test("inference works in for-comprehension") {
+    sealed abstract class AppError
+    case object Error1 extends AppError
+    case object Error2 extends AppError
+
+    val either1: Id[Either[Error1.type , String]] = Right("hi").pure[Id]
+    val either2: Id[Either[Error2.type , String]] = Right("bye").pure[Id]
+
+    for {
+      s1 <- EitherT(either1)
+      s2 <- EitherT[Id, AppError, String](either2)
+    } yield s1 ++ s2
+
+    for {
+      s1 <- EitherT(either1)
+      s2 <- EitherT.right[AppError]("1".pure[Id])
+    } yield s1 ++ s2
+
+    for {
+      s1 <- EitherT(either1)
+      s2 <- EitherT.left[String](Error1.pure[Id])
+    } yield s1 ++ s2
+
+    for {
+      s1 <- EitherT(either1)
+      s2 <- EitherT.pure[Id, AppError]("1")
+    } yield s1 ++ s2
```

Comment:
@kailuowang hi sorry for the late reply. These tests are just to prevent regressions like the one I was going to introduce in #1627 so for the sake of completion I covered the different use cases...